Rlabkey/R/labkey.rstudio.R

```diff
@@ -0,0 +1,194 @@
+        if (is.null(folderPath) || folderPath == "NULL")
+        {
+            return("Unable to determine report folderPath. Skip saving source to LabKey.")
+        }
+    }
+
+    ## normalize the folder path
+    folderPath <- encodeFolderPath(folderPath)
+
+    baseUrl=labkey.getBaseUrl(NULL)
+
+    ## check valid report
+    url <- paste(baseUrl, "rstudio", folderPath, "ValidateRStudioReport.api", sep="")
+    params <- list(entityId=reportEntityId)
+    response <- labkey.post(url, toJSON(params, auto_unbox=TRUE))
+    lkResult <- (fromJSON(response))
+
+    if (lkResult$isValid == TRUE)
+    {
+        localLastModified <- labkey.rstudio.getSavedProp("lastModified")
+        doSave <- TRUE
+        if (localLastModified == lkResult$lastModified)
+        {
+            if (useWarning == TRUE)
+                doSave <- .rs.api.showQuestion("Save to LabKey?", "Do you want to update report content to LabKey Server?")
+        }
+        else
+        {
+            doSave <- .rs.api.showQuestion("Save to LabKey? (Potential conflicting edit)", "The report source was modified in LabKey Server and the content might have diverged from local copy. Do you want to save local changes to LabKey Server?")
+        }
+        if (!doSave)
+        {
+            return("Skipped saving updated source to LabKey server");
+        }
+        url <- paste(baseUrl, "rstudio", folderPath, "SaveRReportContent.api", sep="")
+
+        if (!file.exists(reportFilename))
+        {
+             stop (paste("File doesn't exist: ", reportFilename))
+        }
+        script <- readChar(reportFilename, file.info(reportFilename)$size)
+
+        params <- list(entityId=reportEntityId, runScript=script)
+        response <- labkey.post(url, toJSON(params, auto_unbox=TRUE))
+
+        result <- (fromJSON(response))
+
+        if (result$success == TRUE)
+        {
+            labkey.rstudio.updateProp("lastModified", result$lastModified)
+            return("Successfully updated report source to LabKey Server.")
+        }
+        else
+            warning(result$errorMsg)
+    }
+    else
+    {
+        warning("Failed to update report source to LabKey server. Report doesn't exist.")
+    }
+}
+
+## Read property value form prop.txt
+##
+labkey.rstudio.getSavedProp <- function(propName)
+{
+    propsFilepath <- "props.JSON"
+    if (!file.exists(propsFilepath))
+        return (NULL)
+    props <- fromJSON(txt = propsFilepath)
+    return (props[propName])
+}
+
+## Update property value to prop.txt
+##
+labkey.rstudio.updateProp <- function(propName, propValue)
+{
+    propsFilepath <- "props.JSON"
+    if (!file.exists(propsFilepath))
+        props <- list()
+    else
+        props <- fromJSON(txt = propsFilepath)
+    props[propName] = propValue
+    write(toJSON(props), file=propsFilepath)
+}
+
+## initialize a RStudio session for rlabkey
+##
+labkey.rstudio.isInitialized <- function()
+{
+    return (!is.null(.lkdefaults[["baseUrl"]]))
```

Comment:
When would this helper get called? seems like it should maybe be called from in the labkey.rstudio.initReport function before it does labkey.setDefaults or something like that. 

I don't think I know the scenario well enough, but in addition to checking for the baseUrl should this check for something like that the current working directory is the LabKeyReports dir?

Reply:
This is actually called in R prolog that we generate for RStudio reports. RStudio uses a new R session when user do certain actions, such as knit and run shiny. The new R session doesn't have rlabkey session so as a fallback we get labkey.data and set rlabkey session from local info. prolog actually won't be sourced if working directory is report dir (since source("prolog.R") is in the report that's run, so can skip that.